### ISSUE TITLE
investigate missing HTML elements in metadata

### ISSUE THREAD

**Issue by billreed63** [maqetta/maqetta#398]
Hi Bill,
I see you added metadata to support &lt;b> and &lt;i>. What about &lt;u>, &lt;strike>, &lt;code>,  etc? Should we create an issue to investigate what HTML elements we need to add, and then add those elements?

Ooh, this reminds me:  I see that these are "hidden" widgets, yet they still declare "inlineEdit".  We should get rid of that in all widgets in that file (including those you didn't add).

Bill, can you handle that also?  I'd do it myself, but I'm tied up with other things and don't have time to test the update....

Javier Pedemonte

**Comment by JonFerraiolo** [comment]
@jhpedemonte

Also need to do new HTML5 tags. See http://www.w3schools.com/html5/html5_new_elements.asp. Already had a user requesting <video>. That's probably the highest priority of the HTML5 elements. The new elements for structured documents are low priority - we are doing apps, not docs. The new web forms stuff is also lower priority - way lower priority than the existing form elements: <input>, <select>, etc.

**Comment by jhpedemonte** [comment]
For M4, we should handle some of the HTML5 widgets as mentioned in the previous comment (**input, select, button**, etc).

**Comment by JonFerraiolo** [comment]
Reassigning to @aerwin , particularly to look at INPUT, SELECT, BUTTON and other HTML form elements

**Comment by aerwin** [comment]
The "speclet" for this work can be found on the [HTML Tags in Widget Palette](https://github.com/maqetta/maqetta/wiki/HTML-Tags-in-Widget-Palette) page in the wiki.

**Comment by JonFerraiolo** [comment]
That page was very helpful! I reviewed the above link, cogitated, and formed some recommendations:
- In the Widget palette, I would _not_ include any of the (deprecated?) old-style formatting tags such as CODE, B, I, U, S, SUB, SUP, or the more modern versions STRONG or EM. Similarly, I would not include any of the "list" elements: LI, UL, OL, DL, DT, DD. Instead, my thinking is that if developers want to do intra-paragraph styling or lists, then use the rich text editor widget or enter HTML markup within the SmartInput (e.g., "Don't forget to do &lt;em&gt;this&lt;/em&gt;"), which will cause SmartInput to show the radio button options for Plain text vs HTML markup, or type markup into source view. Putting all of these tags in the widget palette is likely to confuse more than help, and will make the list of widgets (already too large for people to navigate through) even larger. I'm thinking that the widget palette should be restricted to either containers (includes DIVs and SPANs) or UI components.
- I would not include elements that are used only to provide detailed data for a parent control: LEGEND, OPTGROUP, OPTION, and then all of the table elements: TBODY, TR, TD, TH, COLGROUP, COL. Instead, I would provide a nice GUI for that HTML element such as SmartInput where the user enter the data needed for these "sub-elements"
- New elements that I would add include BR, BUTTON, FIELDSET, FORM, INPUT, OBJECT, P, VIDEO (both of which from a UI perspective would work like IMG and IFRAME - present a file picker)
- According to QuicksMode, only Opera supports DATALIST
- I'm assuming we can postpone support for some of the new HTML5 UI widgets such as MENU, METER and OUTPUT

**Comment by aerwin** [comment]
Thanks for the thorough review!
- Those "old-style" formatting tags are not deprecated as far as I can tell, but I'll _not_ include them. A lot of the meta-data already exists in the master branch, so I'll clean that up as part of this work.
- For the "nice GUI" to handle FIELDSET, OPTION, etc. Is that something you'd like to see as part of this work in this sprint?
- I'll include  BR, BUTTON, FIELDSET, FORM, INPUT, OBJECT, P, VIDEO.
- As for DATALIST, it's also supported by Firefox. But, are you suggesting we wait it's more widely supported? If we do it, it's also one that begs for a nice GUI (both to edit contents of the DATALIST itself and associate it with an INPUT field).
- I'll plan on postponing  MENU, METER and OUTPUT

**Comment by jhpedemonte** [comment]
Agreed with most of what Jon said.
- I would include AUDIO in addition to VIDEO.  While the latter is more popular, there were a couple (?) people who mentioned those tags when Maqetta was first released publicly.
- The new way of laying out a page in HTML5 includes ARTICLE, ASIDE, HEADER, FOOTER, NAV, SECTION, etc.  Might be good to include those also.  Maybe as a second pass.  See https://developer.mozilla.org/en/Sections_and_Outlines_of_an_HTML5_document.
- We should eventually include CANVAS. But that carries other baggage with it, so it will be more involved.  We should split that off into a separate bug.
- Do we want to compact the Headings (H1...H6) into one "widget" and handle the differences through smart input?  Would help clean up the widget palette.
- We should move RICHTEXT and STICKYNOTE to a separate category (i.e. "Maqetta"), since they are not HTML elements.

**Comment by JonFerraiolo** [comment]
Responding to questions above:
- AUDIO is fine
- I would be in no hurry with the semantic HTML elements like ARTICLE. I can't see many of our users needed those elements in short-term
- CANVAS is problematic because it is just a tag with an ID. Its content has to be created via JavaScript. 
  Other comments later....

**Comment by aerwin** [comment]
Thanks, guys... per Javier's question about headings, I was debating whether we need at all if philosophy is to send user to Rich Text Editor for formatting (bold, italic, etc.) and list building. I notice you can also do headings in there. Thoughts?

**Comment by aerwin** [comment]
Some new issues created for tracking and to allow new threads of discussion if necessary:
- #1590 for adding HTML5 semantic/layout elements
- #1591 for CANVAS support
- #1592 for lower priority HTML5 UI controls (like MENU, METER and OUTPUT)

**Comment by jhpedemonte** [comment]
Well, I think it would be best to keep the heading tags around.  They are part of the HTML5 document structure, along with the ones I mentioned in my 2nd bullet above.

**Comment by aerwin** [comment]
The speclet (still in draft state) has been updated based on feedback: [HTML Tags in Widget Palette](https://github.com/maqetta/maqetta/wiki/HTML-Tags-in-Widget-Palette)

**Comment by JonFerraiolo** [comment]
I would keep the headings as a special case, even if the other "formatting" elements are left off. The reason is that UIs often use H1-H3 for parts of the UI, such as titles blocks above a bunch of widgets, or subheadings in a flow of widgets (e.g., heading block, then a special widget for a blog comment, then heading block, then blog comment widget, etc). The intra-paragraph formatting elements aren't used in this way (or at least quite a bit less).

**Comment by aerwin** [comment]
@jhpedemonte, @JonFerraiolo 

Hey, Guys... per discussion this morning, I've pushed my local branch ("398_html_widgets") to the central repository for review. If you guys could take a look, I'd appreciate it.

Also, the speclet has been updated in the wiki at [HTML Tags in Widget Palette](https://github.com/maqetta/maqetta/wiki/HTML-Tags-in-Widget-Palette) if you wanted to take one more look at that.

Thanks!

**Comment by JonFerraiolo** [comment]
Great job! Comments after a quick review of the branch:
- I added AUDIO and my goodness some UI appears in the HTML file. I did not know the browsers added default UI for AUDIO. Wow.
- We should include a small bit of audio and video in the sample_data folder of the default workspace. I created bug #1610
- BUTTON has a 'value' attribute. What does it do? Get rid of it?
- I like the way FIELDSET and LEGEND work in general, but I found a problem which maybe should become a separate bug. If I select the FIELDSET, set border to something (1px solid black), at the end of the operation, the LEGEND gets selected, versus preserving the selection on the FIELDSET.
- I think we should drop EMBED. EMBED a deprecated element (vs OBJECT) and in almost all cases where you might use EMBED, you can use OBJECT instead. But I don't think we should include OBJECT either because the primary W3C usage for OBJECT has been referencing SVG files, and all browsers now support IMG pointing to SVG. Given how the browsers have advanced and old quirks mode isn't as critical as before and how browsers are moving away from plugins in general, EMBED and OBJECT usage should shrink to zero, and if a user really needs it, he can type the markup into source view. (But I can be talked into including OBJECT.)
- For INPUT, I don't know what placeholder, src and step are. If I don't know these things, I expect most users won't know. I suggest dropping unless there is a good argument to keep. Advanced users can always type these attributes into source view.
- The text labels in the widget-specific section have first character as uppercase. I think we should literally match the official attribute name, which means all lower case (e.g., 'required' instead of 'Required')
- Widget palette shows LEGEND, OPTGROUP and OPTION. There is a way to prevent these widgets from appearing in Widget palette. @billreed63 knows how.
- We need SmartInput for SELECT. Also, shouldn't there be a value property under widget-specific?

**Comment by aerwin** [comment]
Thanks for the thorough review... some responses...
- For AUDIO, I set the "controls" attribute on by default which is why that shows up. Otherwise, AUDIO is invisible.
- Agreed on #1610.
- For BUTTON, yes, it has a "value" attribute which represents the value to submit to the server when pressed. Would you still vote to get rid of it?
- I'll look into the FIELDSET/LEGEND border thing or open a new issue.
- EMBED is actually what's recommended in HTML5. It's been undeprecated and I get the sense it's preferred over OBJECT (especially now that AUDIO and VIDEO exist). Here's info on the tag: [http://www.w3schools.com/html5/tag_embed.asp].... Do we need a discussion on this?
- Yeah, INPUT bothered me a bit from an overall usability perspective...  It's a weird case in that the applicable attributes change significantly based on the "type" attribute. [http://www.w3schools.com/html5/tag_input.asp]
  - placeholder is fairly important in my opinion as it "specifies a short hint that describes the expected value of an INPUT element." At run time, this is the grayed out text that is shown when text field is empty. This can be used to give the end user an example of what they need to enter in the text field. So, I'd vote to keep this one.
  - src is applicable for type "image" and "specifies the URL of the image to use as a submit button." If we remove this one, then we should probably also remove "image" from the list of valid types. Thoughts?
  - step is applicable for type "number" and "specifies the legal number intervals for an input field." I don't have a problem removing that one.
- For the attribute labels, I can change all of those to match the attribute names... I was following some precedent that seemed to be there to give more "human readable" labels. E.g., the link tag that was already present labeled the "src" attribute as "URL".
- For SELECT/OPTGROUP/OPTION and FIELDSET/LEGEND groupings of elements... I know how to hide them, but I thought we had agreed to deprioritize smart editing for those elements (and their parents). So, right now dragging from the palette is only way to do it. I had already opened #1602 to track the work for additional smart editing. Would you like that to be prioritzed for this sprint?
- For value attribute on SELECT... there is no "value" attribute [http://www.w3schools.com/html5/tag_select.asp]. But, there are atrributes on OPTION for "selected" and "value" [http://www.w3schools.com/html5/tag_option.asp].

**Comment by jhpedemonte** [comment]
See commit 513030772447082e4a8bfff5c8931a080d53bd9b.

**Comment by JonFerraiolo** [comment]
Sorry for slowness in responding. 
- Regarding BUTTON and 'value', I would remove it. Will cause more confusion that value to typical user.
- I did not know that EMBED was undeprecated! Now I'm thinking we should leave it in. I suspect there will be users who want to embed SWF files. But if we put it in, we need something in the default workspace that can be the target of the EMBED element. Does EMBED work with SVG files or IMG files?
- OK, let's keep placeholder and src on INPUT
- Yes, let's match attribute names. IMO, the A element should have an href property instead of URL. (In fact, you want to change that?)
- We don't need SmartInput for SELECT/OPTGROUP/OPTION for this _sprint_ but would be very good to have for this _milestone_ (meaning M5). OK to leave those elements in for the sprint, but if possible I'd like to remove them before shipping M5.
- Regarding 'value' on SELECT, OK, no 'value' property.

**Comment by jhpedemonte** [comment]
Regarding the use of `unconstrained` in widget properties in the OAM:

The [OpenAjax Metadata spec](http://www.openajax.org/member/wiki/OpenAjax_Metadata_1.0_Specification_Widget_Metadata) shows that `unconstrained` is an attribute on the `options` property (notice the 's'), not `option`.

Now, the spec is based on an XML syntax.  However, we use a JSON represenation (never ratified under OpenAjax).

I don't always remember the particulars of our JSON representation, so I look to the converter I wrote, which can be found in _utils/oamXml2Json.js_ -- it's a Node.js script.

So first, I created a _oam.xml_ file for the `<a>` (based on what you've already written, Tony:

``` xml
<widget id="http://www.w3.org/html/a"
        name="html.a"
        spec="1.0"
        version="1.0">
    <properties>
        <property name="href" datatype="string" format="url"/>
        <property name="target" datatype="string" defaultValue="">
            <options unconstrained="false">
                <option value=""/>
                <option value="top"/>
                <option value="bottom"/>
            </options>
        </property>
    </properties>
    <content>
        <![CDATA[
            <a></a>
        ]]>
    </content>
</widget>
```

Notice how I followed the spec and added an `options` child to `property` and put `unconstrained` as an attribute on that.

Then, we run that through the converter:

```
node oamXml2Json.js a_oam.xml
```

and we get the JSON representation:

``` javascript
{
    "id": "http://www.w3.org/html/a",
    "name": "html.a",
    "spec": "1.0",
    "version": "1.0",
    "property": {
        "href": {
            "datatype": "string",
            "format": "url"
        },
        "target": {
            "datatype": "string",
            "defaultValue": "",
            "option": [
                {
                    "value": "",
                    "$parent_unconstrained": "false"
                },
                {
                    "value": "top",
                    "$parent_unconstrained": "false"
                },
                {
                    "value": "bottom",
                    "$parent_unconstrained": "false"
                }
            ]
        }
    },
    "content": "<a></a>"
}
```

So here you see how the `unconstrained` property is converted to JSON.

**Notes:**
1. You might need to change some code to look for the right property.
2. Now that I look at it, I don't particularly like how this gets converted.  I'll have to think of something better.  Please feel free to propose something better.

**Comment by aerwin** [comment]
As for `unconstrained`, I had just put this reply in the review for commit 513030772447082e4a8bfff5c8931a080d53bd9b where you had mentioned the schema:

---

The `unconstrained` attribute is necessary because in addition to (_self, _top, etc.) you can specify the id of any iframe    [http://www.w3schools.com/html5/att_a_target.asp]).

In WidgetProperties.js, it was looking for `unconstrainted` (notice the misspelling) on the property itself. So, I had to change the existing code to spell it correctly so the condition would pass.

BUT, as I look at the link you sent, you are correct that the schema calls for `unconstrained` to be on `options`... 

BUT, interestingly, in studying the schema further, it also says that `option` is only supposed to contain one item. But, no where in any of our OAM metadata files do I see anyone using `options` for multiple items. So, I think most of our OAM metadata files are violating the schema.

Do we want a separate issue to track making all of the OAM metadata files across Maqetta to conform to the schema (e.g., use `options` instead of `option`) and properly handle `unconstrained` as part of `options`?

**Comment by aerwin** [comment]
@jhpedemonte opened #1621 for the options/unconstrained discussion.

**Comment by aerwin** [comment]
@jhpedemonte @JonFerraiolo Made changes based review of previous commit. I'd appreciate it if you could review the latest. Hopefully, most of the issues have been addressed (and those that haven't been are associated with separately opened issues).
